gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -0,0 +1,449 @@
+                              self._process_header_lines),
+        **kwargs)
+
+    self._text_lines = text_source.read_records(self._file_name,
+                                                range_tracker)
+
+  def _process_header_lines(self, header_lines):
+    """ This method will be automatically called by textio._TextSource()."""
+    if self._representative_header_lines:
+      # Replace header lines with given representative header lines.
+      # We need to keep the last line of the header from the file because it
+      # contains the sample IDs, which is unique per file.
+      self._init_with_header(self._representative_header_lines +
+                             header_lines[-1:])
+    else:
+      self._init_with_header(header_lines)
+
+  def next(self):
+    for text_line in self._text_lines:
+      return self._get_record(text_line)
+    raise StopIteration
+
+  def __iter__(self):
+    return self
+
+  def _init_with_header(self, header_lines):
+    # type: (List[str]) -> None
+    """ This method will be called by _process_header_lines()."""
+    raise NotImplementedError
+
+  def _get_record(self, line):
+    # type: (str) -> Variant
+    """ This method will be called by next(), one line at a time."""
+    raise NotImplementedError
+
+class PyvcfParser(VcfParser):
+  """An Iterator for processing a single VCF file using pyvcf."""
+
+  def __init__(self,
+               file_name,  # type: str
+               range_tracker,  # type: range_trackers.OffsetRangeTracker
+               file_pattern,  # type: str
+               compression_type,  # type: str
+               allow_malformed_records,  # type: bool
+               representative_header_lines=None,  # type:  List[str]
+               **kwargs  # type: **str
+              ):
+    # type: (...) -> None
+    super(PyvcfParser, self).__init__(file_name,
+                                      range_tracker,
+                                      file_pattern,
+                                      compression_type,
+                                      allow_malformed_records,
+                                      representative_header_lines,
+                                      **kwargs)
+    # This member will be properly initiated in _init_with_header().
+    self._vcf_reader = None
+
+  def _init_with_header(self, header_lines):
+    self._header_lines = header_lines
+    try:
+      self._vcf_reader = vcf.Reader(fsock=self._line_generator())
+    except SyntaxError as e:
+      raise ValueError(
+          'Invalid VCF header in %s: %s' % (self._file_name, str(e)))
+
+  def _line_generator(self):
+    header_processed = False
+    if not header_processed and self._header_lines:
+      for header in self._header_lines:
+        self._current_line = header
+        yield self._current_line
+      header_processed = True
+    # Infinit loop; VcfParser.next() rasies StopIteration when EOF is reached.
+    while self._current_line:
```

Comment:
Thinking about this more: I think it's better to be explicit and just say `while True:` since we know that the string will never be set to None (see now below). The comment then becomes something like `Continue to process the next line indefinitely. The next line is set inside _get_record and this method is indirectly called in get_record`.
Also note that this behavior is independent of `VcfParser.next()`; the relationship is really between get_record and this method.

note: the current logic has a bug as it should check with `is not None`, otherwise, we'd terminate the pipeline if there's an empty line (PyVcf [skips empty lines](https://github.com/jamescasbon/PyVCF/blob/master/vcf/parser.py#L280)). I think we should add an integration test with an empty line to make sure this continues to work as we make changes. Could you please change, say, valid-4.2.vcf and add an empty line in the middle of the file?